### PR TITLE
fix(932105): update data list

### LINKED ISSUE #2672: update word list for rule 932105 (RCE Unix command injection part 2/4)
Solution and commands should be similar to unix-shell.data.

### PR DESCRIPTION
**Pull request by fzipi**
Signed-off-by: Felipe Zipitria <felipe.zipitria@owasp.org>

Updates the second part of the regexp using unix-shell.data commands.

Fixes #2672.

### CODE REVIEW

File: util/regexp-assemble/data/932105.data [hunk 1]
```diff
@@ -94,6 +94,7 @@ $\(\(
   ##!> cmdline unix
 
 ##! words starting with a-n are in 932100.data
```
Comment (theseion) [inline util/regexp-assemble/data/932105.data:96]: Not true... `m` and `n` are in this file. Maybe we need to move some over here or vice-versa.

File: util/regexp-assemble/data/932105.data [hunk 2]
```diff
@@ -244,35 +295,58 @@ top
...
+wireshark
+wish
 xargs
+xelatex
+xetex
+xmodmap
+xmore
+xpad
 xterm
+xxd
```
Comment (theseion) [inline util/regexp-assemble/data/932105.data:347]: Duplicate

File: util/regexp-assemble/data/932105.data [hunk 3]
```diff
@@ -244,35 +295,58 @@ top
...
 xxd@
+xz
```
Comment (theseion) [inline util/regexp-assemble/data/932105.data:349]: Duplicate

**Review by theseion — CHANGES_REQUESTED** [review]
I found two duplicates, there might be more. The use of `@` is also pretty much a wild guess. Can we maybe come up with a rational? Maybe even move those to a separate rule, or use two separate blocks in the same file?

File: util/regexp-assemble/data/932105.data (hunk 1 above)
Comment (fzipi) [reply util/regexp-assemble/data/932105.data:96]: The comment was wrong. a-l are in that file. Updated comment.

File: util/regexp-assemble/data/932105.data (hunk 2 above)
Comment (fzipi) [reply util/regexp-assemble/data/932105.data:347]: Removed

File: util/regexp-assemble/data/932105.data (hunk 3 above)
Comment (fzipi) [reply util/regexp-assemble/data/932105.data:349]: Removed

**Comment by fzipi** [comment]
I've added the comments, but before pushed wanted to add a version list to python/php/ruby/irb and then I'm stuck. Will update soon.

**Comment by fzipi** [comment]
Looks like there is a problem with the regexp generation. We are tracking this with @theseion .

**Comment by fzipi** [comment]
@theseion I think this is ready to go.

**Comment by RedXanadu** [comment]
@fzipi It shows that the file `932100.data` has been updated but rule 932100 and its regular expression do not appear to have changed. Is that intentional? Or maybe I'm missing something.

**Comment by RedXanadu** [comment]
This regexp-assemble stuff is very clever and very, very complicated…

Why do we need to pick out "python3.10" etc. if we are already matching on just "python"?

Also, do we need to worry about evasions like `python3.1""0` ? Because the `cmdline unix` processor doesn't pick that up.

**Comment by RedXanadu** [comment]
I think so, yes. This needs tweaking like in #2590 , and maybe some tests to check the anti-evasion is working.

I can take a look next week.

**Comment by fzipi** [comment]
@theseion When you have a chance to take a second look...

**Comment by RedXanadu** [comment]
Sorry, I haven't had time to look at this one yet. It's still on my todo list.

**Comment by dune73** [comment]
Thank you for this PR and the fixes @fzipi.

I have taken a closer look and get the feeling the following terms are attracting false positives. Adding the `@` token might prevent that to a certain extent:

```
mail
pry
snap
task
wish
```

@theseion mentiones a double `@` in his review. I think this is now gone. Therefore I believe his requested changes are implemented and can be overridden. But please check yourself @fzipi.

File: util/regexp-assemble/data/932105.data [hunk 4]
```diff
@@ -91,9 +91,10 @@ $\(\(
 (?:[\?\*\[\]\(\)\-\|+\w'\"\./\x5c]+/)?[\x5c'\"]*
 ##!=>
 
-  ##!> cmdline unix
+##!> cmdline unix
```
Comment (theseion) [inline util/regexp-assemble/data/932105.data:94]: Should be indented, it's nested within `assemble`.

File: util/regexp-assemble/data/932105.data [hunk 5]
```diff
@@ -131,69 +144,82 @@ nroff
 nsenter
 nslookup
 nstat
+octave@
+od
 onintr
 openssl
+openvpn
```
Comment (theseion) [inline util/regexp-assemble/data/932105.data:151]: FP? Maybe add `@`.

File: util/regexp-assemble/data/932105.data [hunk 6]
```diff
@@ -131,69 +144,82 @@ nroff
...
+openvt
+opkg
 passwd
+paste@
 patch@
+pax
```
Comment (theseion) [inline util/regexp-assemble/data/932105.data:157]: FP? Maybe add `@`.

File: util/regexp-assemble/data/932105.data [hunk 7]
```diff
@@ -131,69 +144,82 @@ nroff
...
 pkexec
 pkg
 pkg_info
 pkginfo
 pkill
 popd
+pr@
 printenv
 printf@
+pry
```
Comment (theseion) [inline util/regexp-assemble/data/932105.data:183]: FP? Maybe add `@`.

File: util/regexp-assemble/data/932105.data [hunk 8]
```diff
@@ -131,69 +144,82 @@ nroff
...
-ruby18
-ruby19
-ruby20
-ruby21
-ruby22
 run-mailcap
 run-parts
 rview
 rvim
+sash
```
Comment (theseion) [inline util/regexp-assemble/data/932105.data:222]: FP? Maybe add `@`.

File: util/regexp-assemble/data/932105.data [hunk 9]
```diff
@@ -211,14 +237,23 @@ sftp
 sg@
 sh.distrib
 sh@
+shuf
 shutdown
 sleep
+slsh
+smbclient
+snap
```
Comment (theseion) [inline util/regexp-assemble/data/932105.data:245]: FP? Maybe add `@`.

File: util/regexp-assemble/data/932105.data [hunk 10]
```diff
@@ -227,15 +262,25 @@ su@
 sudo
 svn
 sysctl
+systemctl
+systemd-resolve
+tac
 tail
 tailf
 tar@
+task
```
Comment (theseion) [inline util/regexp-assemble/data/932105.data:271]: FP? Maybe add `@`.

File: util/regexp-assemble/data/932105.data [hunk 11]
```diff
@@ -244,35 +289,55 @@ top
...
+wc
 wget
-##! 'who' causes many FP, so require whitespace.
-##! Unfortunately, without whitespace it may still leak some info.
+whiptail
 who@
 whoami
+whois
+wireshark
+wish
```
Comment (theseion) [inline util/regexp-assemble/data/932105.data:332]: FP? Maybe add `@`.

File: util/regexp-assemble/data/932105.data [hunk 12]
```diff
@@ -283,7 +348,10 @@ xzfgrep
 xzgrep
 xzless
 xzmore
+yarn
+yelp
```
Comment (theseion) [inline util/regexp-assemble/data/932105.data:352]: FP? Maybe add `@`.

File: util/regexp-assemble/data/932105.data [hunk 13]
```diff
@@ -295,4 +363,34 @@ zless
 zmore
 zrun
 zsh
+zsoelim
+zypper
+##!<
```
Comment (theseion) [inline util/regexp-assemble/data/932105.data:368]:
```suggestion
  ##!<
```

**Review by theseion — CHANGES_REQUESTED** [review]

File: util/regexp-assemble/data/932105.data (hunk 5 above)
Comment (fzipi) [reply util/regexp-assemble/data/932105.data:151]: On openvpn?

Comment (theseion) [reply util/regexp-assemble/data/932105.data:151]: Not sure. It's just that it's probably quite a common term to find in web documents. But then so are many of the other words in the list.

File: util/regexp-assemble/data/932105.data (hunk 4 above)
Comment (fzipi) [reply util/regexp-assemble/data/932105.data:94]: Noted.

File: util/regexp-assemble/data/932105.data (hunk 5 above)
Comment (fzipi) [reply util/regexp-assemble/data/932105.data:151]: Well, we can talk about this in today's meeting. Do we add `@` to any command that requires parameters?

**Comment by RedXanadu** [comment]
@fzipi After taking a closer look, I don't _think_ that the Unix anti-evasion pattern needs to be added by hand in this example. I think my previous comment can be ignored.

Unlike rule 932150, this rule (932105) isn't looking for a very specific suffix, so I _think_ it's fine.

For example, even though you can successfully evade part of the rule when entering `python3.10` it still matches against "python":

image

So I _think_ it's fine without adding the anti-evasion pattern.

**Comment by fzipi** [comment]
In summary: adding the numbers won't give us added value in this case, so I'm reverting them back to their original positions in the file and removing the specific cases.

**Comment by fzipi** [comment]
:thinking: Wait a minute. What happens with the two words that also are common English words, `python` and `ruby`? Because adding the `@` at the end will change what we are discussing here, and clearly they might benefit from it...

**Comment by theseion** [comment]
What a brainfuck. How about I add another symbol for suffix evasion without space? Then you could add `python~` and `ruby~` and it would mean the same thing as with `@` but without space as a possibility.

**Review by theseion — APPROVED** [review]
LGTM

**Comment by fzipi** [comment]
Thanks. Tests are still failing....

**Comment by fzipi** [comment]
Thanks. At least one of us is awake.

I think I deserve a 🍺 for this bypass then.

**Comment by fzipi** [comment]
Creating a new issue to follow up on test #7.

**Pull request merged**